server/src/main/java/se/fortnox/reactivewizard/server/ConnectionCounter.java

```diff
@@ -0,0 +1,40 @@
+package se.fortnox.reactivewizard.server;
+
+import javax.inject.Singleton;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+@Singleton
+public class ConnectionCounter {
+    private final AtomicLong connections = new AtomicLong(0);
+    private final Semaphore connectionsZero = new Semaphore(1);
+
+    public void increase() {
+        if (connections.getAndIncrement() == 0) {
+            connectionsZero.tryAcquire();
+        }
+    }
+
+    public void decrease() {
+        if (connections.decrementAndGet() == 0) {
```

Comment:
Not a problem that will happen in this PR but what if someone were to call decrease() before an increase() has been called for some reason? Then the lock won't be acquired on the next call to increase() and awaitZero() won't actually wait since it will be able to acquire the lock directly.

Reply:
Not sure what scenario you are thinking of. Is it this?:
decrease()
increase()
increase()
decrease()

That one works just fine...

Reply:
No I was thinking:
decrease()  //this is done when connections = 0, by mistake or some weird reason
increase()
awaitZero()
.
.
decrease()

If the above happens awaitZero() will return immediately even though there is an outstanding request still going on.

Reply:
So your scenario is that the first decrease() is faulty, rather than just being out of order? I guess we cannot support both those scenarios?

With current implementation we get:

**Scenario: first is faulty**
start, count = 0
decrease(), count = -1
increase(), count = 0
decrease(), count = -1
awaitZero() will fail

**Scenario: first is out of order**
start, count = 0
decrease(), count = -1
increase(), count = 0
increase(), count = 1
decrease(), count = 0
awaitZero() passes


If we change to support "faulty" calls:

**Scenario: first is faulty**
start, count = 0
decrease(), count = 0 (instead of -1)
increase(), count = 1
decrease(), count = 0
awaitZero() passes

**Scenario: first is out of order**
start, count = 0
decrease(), count = 0 (should have been -1)
increase(), count = 1
increase(), count = 2
decrease(), count = 1
awaitZero() will fail

So, as I see it, we cannot support both cases,and will have to choose the most likely of these unexpected scenarios. I would say it is more likely to have a race and ending up with calls out of order, than having spontaneus method calls (sunflares?).

Reply:
I was more thinking of a hypothetical case where a user of the class misbehaves. As mentioned, that won't happen with the changes in this PR and probably not ever. I don't think we need to worry about quantum effects spontaneously calling the method :)